### ISSUE TITLE
Stuck in waitOnParentShard after resharding

### ISSUE THREAD

**Issue by tommyblue** [vmware/vmware-go-kcl#72]
Our app reads data from kinesis and we would like to make it quit after a resharding. It runs on ECS and it will be automatically restarted (we use this strategy to mitigate the missing shard stealing feature).

We use dynamodb checkpointing and we noticed that the app using this library can't quit. After some debugging I found out a possible bug, but it feels "strange" so I'm not 100% sure I'm not missing something here 😅 

So, what happens is that, given 1 open shard, when we make a resharding we'll have 2 shards: the original one which is closed and its child, the new shard. The library stops reading the child shard because the parent shard, which is closed, could still contain data that must be read.
The child shard consumer enters the `waitOnParentShard` function and waits `pshard.Checkpoint == chk.SHARD_END` to be true, but nothing sets `chk.SHARD_END` in the dynamodb checkpoint.

We implemented the `ProcessRecords(input *interfaces.ProcessRecordsInput)` function so I guess we could check there if the last read record has the same checkpoint of the shard's `ShardEndingSequenceNumber`. Based on AWS docs `ShardEndingSequenceNumber` is populated only for closed shards and contains the last record's checkpoint in the shard.
`ShardEndingSequenceNumber` is available in `worker.shardStatus`, but it's not public. It could, in case, be added to the input passed to `ProcessRecords`, but it isn't called if there're no records to process, so it's probably not the right path.
So here I'm wondering if there's a missing part in the library or I'm not doing the correct thing.

Looking around in the code I noticed
https://github.com/vmware/vmware-go-kcl/blob/master/clientlibrary/worker/shard-consumer.go#L256 where `getResp.NextShardIterator == nil` is checked to identify a closed shard. Maybe that's the right place to update the chekpoint and set `chk.SHARD_END`.

**Comment by tommyblue** [comment]
A possible solution is to change the code in https://github.com/vmware/vmware-go-kcl/blob/master/clientlibrary/worker/shard-consumer.go#L256 adding `input.Checkpointer.Checkpoint(nil)`:

```go
if getResp.NextShardIterator == nil {
	log.Infof("Shard %s closed", shard.ID)
	if err := input.Checkpointer.Checkpoint(nil); err != nil {
		return err
	}
	shutdownInput := &kcl.ShutdownInput{ShutdownReason: kcl.TERMINATE, Checkpointer: recordCheckpointer}
	sc.recordProcessor.Shutdown(shutdownInput)
	return nil
}
```

**Comment by taojwmware** [comment]
The library has been tested well for resharding scenario. It does not require restart. 
Also, make sure to do checkpointing during shutting down. See example:
https://github.com/vmware/vmware-go-kcl/blob/master/test/worker_test.go#L300

**Comment by tommyblue** [comment]
Thanks for the hint on the shutdown, that's exactly what I was looking for! 👍 

As per the resharding scenario, we decided to shutdown instances after a resharding not for the resharding itself, but to balance the shards between kcl instances.
Some details: we have long living kcl instances reading from the kinesis streams. Those instances live on AWS EC2 spot servers and, hence, those servers can be killed sometimes due to their spot nature. Without shard stealing we found out that instances with longer lives will read from more shards over time, leading to unbalanced works between servers.
So we decided to use `MaxShards` based on `(n.shards / n. instances) + 1` so that all kcl instances will always read from the ~same number of shards over time. 
Since the number of shards changes during a resharding operation, we quit all instances when a resharding happens and then, during the new startup, the correct `MaxShards` value gets calculated and passed to kcl.
